### PR TITLE
Collapse Double Turn

### PR DESCRIPTION
**Pull request by kdiluca**
# Issue

What issue is this PR targeting? If there is no issue that addresses the problem, please open a corresponding issue and link it here.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: test/gurka/test_instructions_uturn.cc [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+class InstructionsUturn : public ::testing::Test {
+protected:
+  static gurka::map map;
+
+  static void SetUpTestSuite() {
+
+    const std::string ascii_map = R"(
+										M N
+										| |
```
Comment (gknisely) [inline test/gurka/test_instructions_uturn.cc:20]: test is messed up again

Comment (kdiluca) [reply test/gurka/test_instructions_uturn.cc]: grr

**Comment by dgearhart** [comment]
Does not fix the example double turn issue and sometimes is too aggressive. Might need to check same names for the source and target edges

**Review by gknisely — CHANGES_REQUESTED** [review]
a few diffs do not look good (i.e., uturns where there should not be any)

 
Screenshot from 2020-11-20 13-36-10

File: src/odin/maneuversbuilder.cc [hunk 2]
```diff
@@ -367,6 +368,37 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
         curr_man = next_man;
         ++next_man;
       }
+      // Combine double L turns or double R turns in short non-internal intersections as u-turns
+      else if (!curr_man->internal_intersection() &&
+               curr_man->travel_mode() == TripLeg_TravelMode::TripLeg_TravelMode_kDrive &&
+               curr_man->length(Options::kilometers) <= (kMaxInternalLength * kKmPerMeter) &&
+               prev_man->HasSimilarNames(&(*next_man), true) && curr_man != next_man &&
+               !next_man->IsDestinationType()) {
+        if ((curr_man->type() == DirectionsLeg_Maneuver_Type_kLeft ||
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:377]: will need to check that the turn degree from the end of the prev man and the begin of the next man is reverse

Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:377]: next_man->set_turn_degree(GetTurnDegree(prev_man->end_heading(), next_man->begin_heading()));

File: src/odin/maneuversbuilder.cc [hunk 3]
```diff
@@ -367,6 +368,37 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
...
+             curr_man->begin_relative_direction() == Maneuver::RelativeDirection::kLeft) &&
+            (next_man->type() == DirectionsLeg_Maneuver_Type_kLeft ||
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:379]: not sure we need both the turn type and relative direction checked

File: src/odin/maneuversbuilder.cc [hunk 4]
```diff
@@ -367,6 +368,37 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
...
+                    next_man->begin_relative_direction() == Maneuver::RelativeDirection::kRight)) {
+          curr_man->set_type(DirectionsLeg_Maneuver_Type_kUturnRight);
+          maneuvers_have_been_combined = true;
+          LOG_TRACE(
+              "+++ Combine: double R turns in short non-internal intersection as ManeuverType=SIMPLE_UTURN_RIGHT +++");
+        }
+        if (maneuvers_have_been_combined)
+          curr_man =
+              CombineDoubleTurnToUturnManeuver(maneuvers, prev_man, curr_man, next_man, is_first_man);
+        if (is_first_man) {
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:397]: can not combine if first man - need to check at beginning of block

Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:402]: therefore, do not need to `is_first_man` pass to CombineDoubleTurnToUturnManeuver

File: src/odin/maneuversbuilder.cc [hunk 5]
```diff
@@ -486,6 +524,26 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
         LOG_TRACE("+++ Combine: obvious maneuver +++");
         next_man = CombineManeuvers(maneuvers, curr_man, next_man);
         maneuvers_have_been_combined = true;
+      }
+      // Combine current short length turn (left or right) with next maneuver that is a kRampStraight
+      // NOTE: This should already be marked internal for OSM data so shouldn't happen for OSM
+      else if (!curr_man->internal_intersection() &&
+               ((curr_man->type() == DirectionsLeg_Maneuver_Type_kLeft ||
+                 curr_man->begin_relative_direction() == Maneuver::RelativeDirection::kLeft) ||
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:532]: not sure about relative check here - how many diffs for this scenario?

File: src/odin/maneuversbuilder.cc [hunk 6]
```diff
@@ -486,6 +524,26 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
...
+                (curr_man->type() == DirectionsLeg_Maneuver_Type_kRight ||
+                 curr_man->begin_relative_direction() == Maneuver::RelativeDirection::kRight)) &&
+               next_man->type() == DirectionsLeg_Maneuver_Type_kRampStraight &&
+               curr_man->length(Options::kilometers) <= (kMaxInternalLength * kKmPerMeter) &&
+               curr_man != next_man && !next_man->IsDestinationType()) {
+        LOG_TRACE(
+            "+++ Combine: current non-internal turn that is very short in length with the next straight ramp maneuver +++");
+        curr_man =
+            CombineDoubleTurnToUturnManeuver(maneuvers, prev_man, curr_man, next_man, is_first_man);
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:541]: need a different combine or possibly reuse another generic one

**Review by dgearhart — CHANGES_REQUESTED** [review]
A quick look at RAD tests found some incorrect diffs

File: src/odin/maneuversbuilder.cc [hunk 7]
```diff
@@ -486,6 +524,26 @@ void ManeuversBuilder::Combine(std::list<Maneuver>& maneuvers) {
         LOG_TRACE("+++ Combine: obvious maneuver +++");
         next_man = CombineManeuvers(maneuvers, curr_man, next_man);
         maneuvers_have_been_combined = true;
+      }
```
Comment (dgearhart) [inline src/odin/maneuversbuilder.cc:527]: Suggestion: split these changes into different PR so the testing/validation will be easier

**Pull request closed without merging**